### DISCUSSION TITLE
Best practises for Python devcontainers with poetry

### DISCUSSION THREAD

**Discussion by SamEdwardes** [devcontainers/community#41] [Q&A]
Most of my Python projects use poetry to manage dependencies. I am wondering if anyone has suggested patterns on how to integrate poetry with devcontainers.

Currently, I have this:

**.devcontainer/devcontainer.json**

```json
{
    "name": "Python 3",
     "build": {
	"dockerfile": "Dockerfile",
	"context": "..",
    }
}
```

**.devcontainer/Dockerfile**

```dockerfile
FROM mcr.microsoft.com/devcontainers/python:0-3.11

USER vscode
WORKDIR /home/vscode

# Install pipx and other dev tools
RUN python3 -m pip install --upgrade pip wheel setuptools \
    && pipx install poetry

# Install the project specific depdencies
COPY poetry.lock ./
COPY pyproject.toml ./
RUN poetry export --with dev --format requirements.txt --output requirements.txt \
    && pip3 install -r requirements.txt
```

I imagine there are better ways to do this though. Does anyone have suggestions? A few things I would like:

- For the docker image to include all of the dependencies
- The docker image to start in the project directory, with a poetry virtual environment pre-created

**Comment by schlich** [comment]
Use the poetry [feature](https://containers.dev/features) (to install dependences) and run `poetry install` in the `postStartCommand` or `postAttachCommand` depending on your needs.  for local caching, make sure virtualenvs.in-project is set to true.

**Reply by SamEdwardes** [reply]
Thank you - I will take a look at this!
